Migration to Webpack (WIP)

HerrVoennchen opened the pull request:
Hi,

Happy Hacktoberfest :)

I have taken a try on issue #156 / #194. This is just a first move since due to the custom build scripts it will take some time. Feel free to look into the changes and may point me to some edges since I’m still exploring the whole package. :)

What’s actual working is build ami and the examples. The size of the library got down from 7MB to 3.3 MB / 1.7MB to 1.3 MB in production. 

 Any feedback is appreciated.

NicolasRannou commented on .babelrc:
```diff
@@ -1,10 +1,25 @@
-    "cjs": {
-      "plugins": ["add-module-exports"]
+    "env": {
+        "cjs": {
```
cjs stuff is legacy, we can remove !

HerrVoennchen replied:
Ok.  I’ll continue to work on this in the evening when the kids are in their beds. :)

NicolasRannou commented:
Great! I'll have a look asap

NicolasRannou commented on package.org.json:
```diff
@@ -0,0 +1,105 @@
+{
```
what is this file for?

HerrVoennchen replied:
Just the original package.json for comparison reasons. When all things are done it can be deleted.

NicolasRannou commented:
I'll make sure to provide more details about the use cases before this evening!

NicolasRannou commented:
We have example and demo just to make sure we can:

* example: run app + ami as one package
* demo: run app + ami.js build package

What is important to be able to do it to have some live reload with `webpack-dev-server` as we usually add/test new feature by adding/updating examples/demos.

Typically I start a dev server targetting a specific example:
```
npm run example exampleID
```
then it rebuilds/reloads each time a file is modified.

Same goes for the demos, the difference being that it just rebuilds ami.js and not the whole thing.

The demo directory have this structure so it can be directly loaded in JS Fiddle for quick demo.

Hope that helps a little bit - nothing is fixed do not hesitate to break things / suggest changes!!

NicolasRannou commented:
Great! Yes I was thinking the same while working on it and maybe that doesn't totally make sense...?

Open to any suggestion to make things more straight forward -

One question, to `babel` only run on AMI and not the node dependencies? Point is we want the auto-reload to run as fast as possible -

In addition, maybe we just need babel to ship things but not in the dev environment? Just thinking out loud and throwing some ideas -

NicolasRannou commented:
BTW, with `webpack` we should be able to know/track down why the `ami.js` build is still so big, correct?

HerrVoennchen commented:
The library itself is not the issue. It's the complex interaction with the examples and demos. I will have a deeper look to these to see, what really happen with these on dev/build and where it can be easier. 

Babel: we want babel(ing) the deps too. With ES6 Imports we can tree shake them too and only bundle what really is used. So one step in this is to make sure there are used/requires get replaced with them.

Also the library currently only targets a node version.  May some minimal Browser versions would be good to tell babel to transpile for these too. Otherwise only Browser Version will work who support the JS Capabilities of the specified Node version.

to see whats in the Bundle you can utilize tools like [webpack-bundle-analyzer](https://www.npmjs.com/package/webpack-bundle-analyzer) to see whats in it and how much space it take. I think out of the box three.js will be quite space eating.

The initial bundling still takes quite some time. When running the dev-server then changes afterwards will be much faster transpiled and bundled. Alternately wie exclude specific dependencies you use and declare them as needed deps on install. (peer?) So the user has to install ami and threejs like other libs do this with e.g. jquery. Your ami package should become much smaller and faster then. At the moment it contains everything in it ami need to run.

NicolasRannou commented:
Sounds all great thats a lot for your help and time!

Just as a quick note, we do not include THREEJS in the bundle as unfortunately it was creating all sort of issues, so the large bundle size is most likely coming from the libraries we use to parse the data, but still I'd expect it to be much smaller!

Can detail the THREEJS issues if you'd like -

Getting the user to install only the packages he needs may be a solution, but we wanted to be it as user friendly as possible -

HerrVoennchen commented:
I analized the bundle. Three.js is part of it since it is referenced/imported in some of the sources. leaving it out will left the bundle dysfunctional (at least as browser-consumable).

I would suggest the following: Since webpack and dev server are working kinda different compared to the previous libs at first step clean the package from everything exept the library itself. Creating a package wich can be uploaded to npm and also be consumed in a browser. the current config at least deliver both these builds.

As a second step the whole example/lesson area should get much simplier. The examples should work right off when the lib is built. Is the copy/merge step really necessary?

(unreflected Idea) May considering to provide a seperate repository for the examples and lessons (which are partly quite complex) which consume the ami lib itself. Unsure how the DX would be like.

NicolasRannou commented:
Can you detail what you mean by "npm" and "browser" version?

> I analyzed the bundle. *Three.js* is part of it since it is referenced/imported in some of the sources. leaving it out will left the bundle dysfunctional (at least as browser-consumable).

Yes that is true I forgot about that - however it should only import a few *Three.j*s modules (Vectors* and Matrix*) so I was hoping it would have small impact on the bundles size -

One of the current requirements of *Ami* is that you must load *Three.js* before to run it in your Web Browser. Tried to get rid of this requirement but it is just not the was *Three.js* works.

There are lot of options there but maybe for later...

> First step clean the package from everything except the library itself. Creating a package which can be uploaded to npm and also be consumed in a browser. the current config at least deliver both these builds.

> As a second step the whole example/lesson area should get much simplier. The examples should work right off when the lib is built. Is the copy/merge step really necessary?


I think I understand, please correct if I am wrong:

1- generate 2 builds (npm and browser)
2- demo/example only target the 2 builds - dev-server should be able to auto reload demos/examples as soon as a new build occurred, right? I like the idea -

copy/merge is not necessary - the only thing to keep in mind is that we currently build all examples to a specific location in order to be able to easily deploy all of them to the github page - maybe that mean an extra webpack command that takes *Ami* umd + *example* and bundles it to a specific location?

Thanks a lot seems like we are almost there!

HerrVoennchen commented:
> Can you detail what you mean by "npm" and "browser" version?

Well, as for now, the now created library bundle can do both with the same file. I made som changes to the webpack config and now a "named" UMD file will be created. This works with node and within browsers.

I commited some code. I started to rebuild the build and deployment scripts. What works:
 
- lib is created for development and production
- examples work out of their folders, (no copying anymore for run and development). for Deployment they will be copied to a dist folder (not yet finished)
- build and dist folder (as defined before) are now merged to a lib folder (with is more or less the default distribution folder for libraries). The dist folder will return as a folder for lib + examples + lession package for deployment but for a npm distribution of ami itself it isn't necessary.

sry, but had only 2-3 h time this weekend for this. Kids+wife are demanding :)

One more word to the examples. The index.html. files are generated from a sample. It is similar to the generation you did in your scripts. It was necessary because for webpack dev server a working index.html page was mandatory to work. Since the examples are accessing the ami sources (with is fine for now, I may change it to the ami bundle afterwards) webpack dev server is bundling them in place (and only in memory, means webpack dev server serves the bundled lib only from RAM, it is not persited to the filesystem unlike webpack.)

let me know, if this is the direction you'd like to have it. I've not tested all examples, but the ones I tested had worked.

regards,

Sebastian

NicolasRannou commented:
Thanks you so much for your progress. I actually think "2-3h" is a lot of time a really appreciate your efforts. :)

All sounds good, especially the babel stuffs that I didn't know about -

If some examples work I am pretty sure all will - maybe make sure to check out this example: https://fnndsc.github.io/ami/#loader_nrrd as it imports the "THREE.VTKLoader" from an external JS file.

Again, thanks a lot for your great work!

HerrVoennchen commented:
almost there. 

I worked over the lessons and examples. Now they work in place incl. hot reload. Because the lessons use the final bundle and not the sources I added a plugin where you can add file(types) that should be watched and page reloaded when some change. Background is that webpack(devServer) only hot reloads the bundle it creates. Which is the demo.js/(example).js in either case. Hence this requires an addon in the Browser which, as far as I know now is only available for Chrome called LiveReload. It automaticly refreshes the page when a file changes that matches the scheme. That way the used ami bundle is watched. A seperate webpack instance is watching the ami sources and the chain is complete.

now the only thing left is the deployment. I will look into this next to see what the dist folder need to consist of for the gh-pages command.

HerrVoennchen commented:
Hi,

I think we have it. I redone the deployment part. On deploy ami gets build and examples will be copied into dist and compiled and all should be ready to upload on gh-pages. (obvisouly I haven't tried this :-D ).

I was able to get rid of the babel-polyfill, the es2015 babel-preset and the add-exports-module babel plugin. The babel-env-preset takes care of all real needed polyfills based on the min-browser specs.
As a Result the package got a little smaller. Now about 2.9MB normal, 1.2 MB Prod-build, just about under 300KB gzipped. (I added a gzipped version on prod build step). If you ditch some older Browser versions the package will shrink even more.

The only thing left is the old "demo" script entry. Since it serves the demo may you tell me it this step is even necessary anymore. Lessons got served from dev-server right in place as well as examples.

What a ride :)

NicolasRannou commented:
Sorry for the late answer I was traveling with no internet connection -

Looks great, I'll test it and let you know if all good!
Last step I can think of is to update the README file!

The previous "demo" was also updating the three.js and ami.js version in all demo.html files, based on the one listed in `package.json` - does that make sense?

https://github.com/FNNDSC/ami/blob/dev/lessons/04/demo.html

HerrVoennchen commented:
No problem, I updated the Readme. There was very little to update, just some details. Besides on yarn you can leave out the run param. just yarn and the build-script name.

Did the demo script only that? I have to look in the old scripts. When you run the gen:index:lesson script, die demo.html files will be updated as well. When there is a need to update the html files run the gen:index scripts and all good. The example and lesson run scripts checks if there are html files and generate some if not existent.

NicolasRannou commented:
Thanks -  I'll go over it and merge it ASAP (currently traveling)

NicolasRannou commented:
Hackfest related issues:
https://github.com/FNNDSC/ami/issues/194

NicolasRannou commented:
Will also close https://github.com/FNNDSC/ami/issues/109

NicolasRannou commented:
@HerrVoennchen I am testing it right now -

When I try to run an example, it seems that after compilation I have to open the console + shift reload the page for it to appear - did you experience that too?

NicolasRannou commented:
```
nico@OSXLAP01821-2:~/work/gitroot/ami (HerrVoennchen-dev *)$ yarn example geometries_slice
```

NicolasRannou commented:
I seem to experience the same behavior with "hot reload" (got to shift reload)

NicolasRannou commented:
same behavior observed for `lesson`:
`yarn run lesson 01`

NicolasRannou commented:
Finally "deploy" looks good, it just misses the " && gh-pages -d dist" command after preparing the `dist` directory

NicolasRannou commented:
Hopefully the "auto" refresh is easy to address and I can merge it right after!

Thanks a lot! Really happy with how it turned out!

HerrVoennchen commented:
Well I tested it again.

examples: work fine at me. The examples are referencing the ami source directly so changes on files the examples imports are resulting on an immediatly reload after webpack has rebundled. Have tested it on Firefox 57 and Chrome. I dropped a console log in the helper.stack.js file. let me know what you tried exactly, I will try this out too.

lessons: It works as expected. Special here is that the lesson source are referencing ami as its final bundle. So Webpack dev server is only watching its created bundle and not the "third party" bundle. The seperate Webpack instance in the background is creating a new ami.js file but the HMR from Webpack dev Server is not aware of this. 
This is where the liveReload plugin for Chrome comes into the game (also available to Firefox up to Version 56. From 57 and Up it is not (yet) compatible with the new AddOn System). It watches all served files (like configured) and reloads if one has changed.

I will try this evening to teach webpack dev server to watch more files that its created bundle but no guarantee that I'll succeed.

I already pushed some small changed. I added the gh-pages command.

NicolasRannou commented:
Thanks for the feedback - I didn't realize live-reload was running as a browser plugin, I thought we were using the npm package to watch the bundle -

That would be great if we could get right of the browser plugin by which ever way! Thanks

I'll Google around too

NicolasRannou commented:
I ran the code without having the live reload chrome plugin, which probably explain why I had to refresh the page manually

HerrVoennchen commented:
Well yes and no. Webpack takes care and hot reloads on the bundle it creates. But not hot reloading files which are part of the app but not Part of the bundle. That’s the plugin for to fill that gap.

NicolasRannou commented:
(on my phone currently so hard to look at the code) - so please correct me if I'm wrong- examples should auto refresh without the live reload plugin but demos are currently expected not to auto refresh without the plugin

NicolasRannou commented:
I'll test it tomorrow on a fresh checkout and report all steps and versions! Thanks

HerrVoennchen commented:
(I’m on the phone too :)). Yes, because the lessons consume ami not es part of the bundle but es part of the whole app. 

You have to separate (on mind) the bundling and watch part with the serving assets part of the dev server (which is a express.js server)

Despite that fact I would always recommend to use ami directly from source like the examples do. With tree shaking only used parts get bundled and keep the file size down.

NicolasRannou commented:
Thx for feedback - so still experiencing same behavior on fresh clone.

no livereload plugin installed on chrome -
```
git clone <>
cd <>
yarn install
yarn example geometries_slice
```

The browser tab opens and seems to be waiting for ressources, and as soon as webpack `webpack: Compiled successfully.` chrome fails to load the asset: `Failed to load resource: net::ERR_CONNECTION_REFUSED`

HerrVoennchen commented:
That’s sounds more like the webpack process failed. Please check the console output. 

I will check out a fresh copy of the repo and try it again. Maybe something is missing but my git doesn’t show anything not checked in.

NicolasRannou commented:
Yes there is nothing suspicious in the webpack log:

```
nico@OSXLAP01821-2:~/work/gitroot/test/ami (dev *)$ yarn example viewers_quadview
yarn example v0.24.5
$ node ./scripts/router.js examples viewers_quadview
LiveReload server running on port 35729
 10% building modules 1/1 modules 0 active                                    e Project is running at http://localhost:8080/
webpack output is served from /
Content not from webpack is served from /Users/nico/work/gitroot/test/ami/examples/viewers_quadview, /Users/nico/work/gitroot/test/ami/lib
404s will fallback to /index.html
 10% building modules 5/8 modules 3 active ...ode_modules/webpack/hot/dev-server.jswebpack: wait until bundle finished: /index.bundle.js
 10% building modules 8/9 modules 1 active ...t/ami/node_modules/webpack/hot/log.jswebpack: wait until bundle finished: /service-worker.js
Hash: 16c8c42e71407d89858b                                                    / Version: webpack 3.7.1
Time: 11378ms
                  Asset     Size  Chunks                    Chunk Names
    viewers_quadview.js  2.77 MB       0  [emitted]  [big]  main
viewers_quadview.js.map   3.5 MB       0  [emitted]         main
  [22] ./src/core/core.utils.js 8.05 kB {0} [built]
  [88] multi (webpack)-dev-server/client?http://localhost:8080 webpack/hot/dev-server ./examples/viewers_quadview/viewers_quadview.js 52 bytes {0} [built]
  [89] (webpack)-dev-server/client?http://localhost:8080 7.23 kB {0} [built]
  [97] ./node_modules/loglevel/lib/loglevel.js 7.74 kB {0} [built]
  [98] (webpack)-dev-server/client/socket.js 1.04 kB {0} [built]
 [130] (webpack)-dev-server/client/overlay.js 3.73 kB {0} [built]
 [135] (webpack)/hot nonrecursive ^\.\/log$ 170 bytes {0} [built]
 [136] (webpack)/hot/dev-server.js 1.61 kB {0} [built]
 [138] ./examples/viewers_quadview/viewers_quadview.js 25.2 kB {0} [built]
 [140] ./src/controls/controls.trackballortho.js 13.9 kB {0} [built]
 [141] ./src/controls/controls.trackball.js 21.2 kB {0} [built]
 [142] ./src/helpers/helpers.localizer.js 6.61 kB {0} [built]
 [146] ./src/helpers/helpers.stack.js 13 kB {0} [built]
 [218] ./src/shaders/shaders.contour.uniform.js 825 bytes {0} [built]
 [220] ./src/shaders/shaders.contour.fragment.js 4.03 kB {0} [built]
    + 206 hidden modules
webpack: Compiled successfully.
```

NicolasRannou commented:
OK so the problem was my version of Chrome (61) - testing on Chrome Canary 64, Safari and Firefox stable properly reload - so that is all all good.

Can we add a note in the README to remember that we need to enable the `livereload` extension on chrome for lessons to auto reload?

Finally it also seems google analytics are incorrect when running the "gen:index:*" scripts.
* lessons, the google key is missing
* examples: no analytics

Can you confirm?

Last thing (I promise), I am also thinking that what would greatly help is a full description of new build system: which tools are being used (some description and references), what operation can we run, which are the steps in the build/dev pipelines, reason for each step.

The reason is that as a medical software, that documentation is required to bring products to the market and that would be a huge help as you are currently working on that, all your thoughts on the subjects are fresh and well organized. Will open a new `hackfest` issue if you are interested into it - Can do it after merging this PR if you'd like - 

Thanks for your patience and thanks a lot for the PR!

NicolasRannou commented:
#205

HerrVoennchen commented:
I agree on documenting the build system. I will put a draft in the next commit.

I have to check my version of Chrome but its the current stable. On Firefox I have the Developer  Edition and the Nightly running.

I'll check on the google analytics. I was thinking of it would only necessary when deploying, not on development. I'll also check the key.

NicolasRannou commented:
Awesome!

About analytics, we need to add in the `examples` before deploying, and in the `demo.html` so when the file is loaded in JSFiddle we get some numbers on that too!

HerrVoennchen commented:
I fixed the google analytics part and put in an initial version of build.md. Beware! English is not my native language :D so it could contain some errors or weird expressions.

May alter the issue to improve this build documentation..

NicolasRannou commented:
Thanks! The build doc is a great start - last thing about it, can we add a bit more details in the build and test sections:

for instance:
build:ami  generates the `ami.js` file. The pipeline to generate the plugin is the following: babel with options <option 1, option2> because <reason for 1, reason for 2>, uglify with those options etc.

and basically same for the minified/zipped version.

and something similar for the test too would be great.

Thanks!

HerrVoennchen commented:
I‘ve written some lines on how webpack is working and also some info on tests an linting.

NicolasRannou commented:
looks good - will test it one last time now-

small note, there is no SASS/SCSS compilation happening right? (it is mentioned in the `build.md`)

HerrVoennchen commented:
You are right. I checked the current config and it's not configured for sass/less. I can change the build.md or, for future use I can add the necessary configurations for sass and less so it can be used out of the box and drop a hint in the build.md anyway.

Because of my other projects I have the needed config changes available.

NicolasRannou commented:
closes #204 #205 #156 #194

HerrVoennchen commented:
thx for merging. Was a fun task to do. If you find any more bugs feel free to contact me directly. I'm happy to fix them.
